Fix: Product Collection block does not display when editing template/post

imanish003 opened the pull request:
### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #47868 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

**Note for Developers:** We were able to reproduce it consistently with "prod" build instead of "dev" build, so I would suggest you to test it on "prod" build. 

**Testing in Product Catalog Template:**
1. From admin go to Appearance > Editor
2. From the design editor go to Templates > Product Catalog (edit)
3. Replace the Products (Beta) block with the new Product Collection block
4. Save the template
5. Now go back to the admin and then back to Appearance > Editor > Templates > Product Catalog (edit)
	- Verify that the Product Collection block is showing a list of products in the Editor. 

**Testing in new post**
1. Create a new post
2. Add Product Collection block to it & choose "Product Catalog" template
3. Save the draft
4. Refresh the post and verify that the Product Collection block displays the products.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Fix: Product Collection block does not display properly when editing template/post

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

imanish003 committed:
Fix issue where Product Collection wasn't showing anything

- Moved queryId logic to a new `useQueryId` custom hook to encapsulate the logic.
- Simplified attribute setting in `useEffect` by using a default attributes value.
- Utilized `fastDeepEqual` to check for attribute equality

These changes fix the issue where the Product Collection block did not display properly in the template editor, ensuring consistent rendering of products.

imanish003 committed:
Add comment

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/product-collection-content.tsx:
```diff
@@ -29,6 +30,42 @@ import InspectorControls from './inspector-controls';
+	// on the same page, updating one will cause the other to be re-inserted.
+	// If we change the ID on init it will trigger a loop as each competes
+	// to set a new queryId and update the sync pattern.
+	// const blockParents = getBlockParentsByBlockName( clientId, 'core/block' );
```
There's some leftover here

imanish003 replied:
Good catch! Removed it in [4550631](https://github.com/woocommerce/woocommerce/pull/47871/commits/45506313e4b54b78377c2721824c854087f51570)

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/product-collection-content.tsx:
```diff
@@ -29,6 +30,42 @@ import InspectorControls from './inspector-controls';
+	// If we change the ID on init it will trigger a loop as each competes
+	// to set a new queryId and update the sync pattern.
+	// const blockParents = getBlockParentsByBlockName( clientId, 'core/block' );
+	const blockParents = useMemo( () => {
```
Is this `useMemo` necessary for a fix or it's just an additional optimization?

imanish003 replied:
It's just for additional optimization & isn't necessary.

kmanijak reviewed (commented):
I suggest including a step in testing steps for developers to test it on `prod` build. I was able to consistently reproduce the issue on `prod` build but not necessarily on `dev` build.

But, this PR fixes the issue for me! So great job Manish! I left some minor comments only. 

Also, do you think we could write an E2E test for it?

imanish003 committed:
Remove leftover comment

imanish003 commented:
> I suggest including a step in testing steps for developers to test it on prod build. I was able to consistently reproduce the issue on prod build but not necessarily on dev build.

Aha! Adding it. Thanks for pointing it out 🙌🏻

> Also, do you think we could write an E2E test for it?

Sure thing. Let me add an E2E test for it. First, I thought about adding it as a follow-up, but as we discussed during the meetup, we should add it with the PR itself. Therefore, I'm going to add it to it now. 🤝

imanish003 committed:
Add e2e tests for Product Collection block visibility after refresh

Added end-to-end tests to verify the visibility of the Product Collection block after page refresh in both Product Archive and Post contexts.

- Added tests to ensure the Product Collection block remains visible after refreshing the template or post editor.
- Updated existing tests to use `BLOCK_LABELS` for better readability and maintainability.

imanish003 commented:
@kmanijak Added E2E tests in [8a55305](https://github.com/woocommerce/woocommerce/pull/47871/commits/8a55305a738c4464c7d7a46707a49c8d8def5d0f). Can you please take a look again?

kmanijak commented on plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts:
```diff
@@ -1151,12 +1154,47 @@ test.describe( 'Product Collection', () => {
+			await pageObject.replaceProductsWithProductCollectionInTemplate(
+				'woocommerce/woocommerce//archive-product'
+			);
+			await expect(
+				editor.canvas.getByLabel( BLOCK_LABELS.productTemplate )
+			).toBeVisible();
+
+			// Refresh the template and verify the block is still visible
+			await page.reload();
+			await expect(
+				editor.canvas.getByLabel( BLOCK_LABELS.productTemplate )
+			).toBeVisible();
+		} );
```
Nit: the selector could be extracted:

```suggestion
            const productTemplate = editor.canvas.getByLabel( BLOCK_LABELS.productTemplate )
			await expect( productTemplate ).toBeVisible();

			// Refresh the template and verify the block is still visible
			await page.reload();
			await expect( productTemplate ).toBeVisible();
```

imanish003 replied:
Good suggestion! Made the changes in [3a6255c](https://github.com/woocommerce/woocommerce/pull/47871/commits/3a6255c34c1580ea96cb1831dba527e4db7c41e1)

kmanijak commented on plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts:
```diff
@@ -1151,12 +1154,47 @@ test.describe( 'Product Collection', () => {
+
+		test( 'In a Post', async ( { page, pageObject, editor } ) => {
+			await pageObject.createNewPostAndInsertBlock();
+			await expect(
+				page.getByLabel( BLOCK_LABELS.productTemplate )
+			).toBeVisible();
+
+			// Refresh the post and verify the block is still visible
+			await editor.saveDraft();
+			await page.reload();
+			await expect(
+				page.getByLabel( BLOCK_LABELS.productTemplate )
+			).toBeVisible();
```
Nit: same as above

```suggestion
            const productTemplate = editor.canvas.getByLabel( BLOCK_LABELS.productTemplate )
			await expect( productTemplate ).toBeVisible();

			// Refresh the post and verify the block is still visible
			await editor.saveDraft();
			await page.reload();
			await expect( productTemplate ).toBeVisible();
```

kmanijak reviewed (approved):
Thanks for adding E2E tests for it! <3 I left some nitpick suggestions but otherwise all good! Approving! 🚢

imanish003 committed:
Improve E2E tests

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into fix/47868-product-collection-block-does-not-display-properly-when-editing-template
